### PR TITLE
Added support for push notification permissions

### PR DESCRIPTION
**Pull request by confile**
Added support for push notification permissions

### CODE REVIEW

**Comment by confile** [comment]
@rsattar please take a look.

**Comment by danipralea** [comment]
it's not merged yet? I can't see the changes

**Comment by rsattar** [comment]
Thanks @confile for writing this up. 

TL;DR I'm going to merge this in (with some changes), but add a warning about using it (I don't recommend recommend using it).

I'm *extremely* hesitant about adding push notifications permissions, because it differs from *ALL* the other permissions-based flows on iOS in significant ways:

1. **You can't reliably know if you have push notifications permissions.** 

In a library like ClusterPrePermissions, you have to use some sort of local "flag", like in `NSUserDefaults`. However this is not great because if you uninstall the app and re-install it, the `NSUserDefaults` are cleared (but your push notifications authorization doesn't change for 24 hours), so your state is lost.

2. **You can't reliably tell when the Push Notification request has finished being made to the user.**

Checking for `UIApplicationDidBecomeActive` is nice, but:
   * It's not guaranteed to fire. 

The user may have explicitly denied permission to the app previously, or you uninstall/reinstall the app (see `1.`). And now you are expecting a dialog to come and go (firing the notification), but it will never fire. Your completion block will not be called, and the app developer will never get a callback about the flow finishing.
   * It can fire prematurely. For example, the user could receive a calendar alert and then dismisses it, before the push notification alert is shown.

--

In my opinion, the most reliable way to do push notification pre-permissions is to be part of the AppDelegate callbacks (`- application:didRegisterUserNotificationSettings:`), and even then you may have to record some NSUserDefaults to know if the USER has asked for permissions before (declaring an intent, like "should-register-always-from-now").

I'm going to merge this in, but with a disclaimer: I don't recommend using it. You should handle "pre-permission" for push notifications yourself. If folks are having issues with this change, then I might remove push notifications support in the future.

**Comment by confile** [comment]
@rsattar Interesting though could you please be more specific how you would do it in ``application:didRegisterUserNotificationSettings``?
